internal/stats/course.go

```diff
@@ -143,31 +79,53 @@ func AsyncStoreCourseMetric(metric *CourseMetric) {
 }
 
 func AsyncStoreCourseGradingTime(startTime timestamp.Timestamp, endTime timestamp.Timestamp, courseID string, assignmentID string, userEmail string) {
-	metric := &CourseMetric{
-		BaseMetric: BaseMetric{
-			Timestamp: startTime,
-		},
-		Type:         CourseMetricTypeGradingTime,
-		CourseID:     courseID,
-		AssignmentID: assignmentID,
-		UserEmail:    userEmail,
-		Value:        uint64((endTime - startTime).ToMSecs()),
+	attributes := map[string]any{
+		TYPE_KEY:  CourseMetricTypeGradingTime,
+		VALUE_KEY: uint64((endTime - startTime).ToMSecs()),
+	}
+
+	if userEmail != "" {
+		attributes[USER_EMAIL_KEY] = userEmail
+	}
+
+	if courseID != "" {
+		attributes[COURSE_ID_KEY] = courseID
+	}
+
+	if assignmentID != "" {
+		attributes[ASSIGNMENT_ID_KEY] = assignmentID
+	}
```

Comment:
I've seen this almost exact code elsewhere.
Maybe you can abstract it out to a constructor for BaseMetric.